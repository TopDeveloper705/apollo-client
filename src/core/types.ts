import { DocumentNode, GraphQLError } from 'graphql';

import { ApolloCache } from '../cache';
import { FetchResult } from '../link/core';
import { ApolloError } from '../errors';
import { QueryInfo } from './QueryInfo';
import { NetworkStatus } from './networkStatus';
import { Resolver } from './LocalState';
import { ObservableQuery } from './ObservableQuery';
import { Cache } from '../cache';

export { TypedDocumentNode } from '@graphql-typed-document-node/core';

export type DefaultContext = Record<string, any>;

export type QueryListener = (queryInfo: QueryInfo) => void;

export type OnQueryUpdated<TResult> = (
  observableQuery: ObservableQuery<any>,
  diff: Cache.DiffResult<any>,
  lastDiff: Cache.DiffResult<any> | undefined,
) => boolean | TResult;

export type PromiseResult<T> =
  T extends PromiseLike<infer U> ? U : T;

export type RefetchQueryDescriptor = string | PureQueryOptions;
export type RefetchQueryDescription = RefetchQueryDescriptor[];

// Used by ApolloClient["refetchQueries"]
// TODO Improve documentation comments for this public type.
export interface RefetchQueriesOptions<
  TCache extends ApolloCache<any>,
  TResult,
> {
  updateCache?: (cache: TCache) => void;
  // Although you can pass PureQueryOptions objects in addition to strings in
  // the refetchQueries array for a mutation, the client.refetchQueries method
  // deliberately discourages passing PureQueryOptions, by restricting the
  // public type of the options.include array to string[] (just query names).
  include?: string[];
  optimistic?: boolean;
  // If no onQueryUpdated function is provided, any queries affected by the
  // updateCache function or included in the options.include array will be
  // refetched by default. Passing null instead of undefined disables this
  // default refetching behavior for affected queries, though included queries
  // will still be refetched.
  onQueryUpdated?: OnQueryUpdated<TResult> | null;
}

// Used by QueryManager["refetchQueries"]
export interface InternalRefetchQueriesOptions<
  TCache extends ApolloCache<any>,
  TResult,
> extends Omit<RefetchQueriesOptions<TCache, TResult>, "include"> {
  // Just like the refetchQueries array for a mutation, allowing both strings
  // and PureQueryOptions objects.
  include?: RefetchQueryDescription;
  // This part of the API is a (useful) implementation detail, but need not be
  // exposed in the public client.refetchQueries API (above).
  removeOptimistic?: string;
}

export type OperationVariables = Record<string, any>;

export type PureQueryOptions = {
  query: DocumentNode;
  variables?: { [key: string]: any };
  context?: any;
};

export type ApolloQueryResult<T> = {
  data: T;
  errors?: ReadonlyArray<GraphQLError>;
  error?: ApolloError;
  loading: boolean;
  networkStatus: NetworkStatus;
  // If result.data was read from the cache with missing fields,
  // result.partial will be true. Otherwise, result.partial will be falsy
  // (usually because the property is absent from the result object).
  partial?: boolean;
};

// This is part of the public API, people write these functions in `updateQueries`.
export type MutationQueryReducer<T> = (
  previousResult: Record<string, any>,
  options: {
    mutationResult: FetchResult<T>;
    queryName: string | undefined;
    queryVariables: Record<string, any>;
  },
) => Record<string, any>;

export type MutationQueryReducersMap<T = { [key: string]: any }> = {
  [queryName: string]: MutationQueryReducer<T>;
};

export type MutationUpdaterFunction<
  TData,
  TVariables,
  TContext,
  TCache extends ApolloCache<any>
> = (
  cache: TCache,
  result: Omit<FetchResult<TData>, 'context'>,
  options: {
    context?: TContext,
    variables?: TVariables,
  },
) => void;
export interface Resolvers {
  [key: string]: {
    [ field: string ]: Resolver;
  };
}
