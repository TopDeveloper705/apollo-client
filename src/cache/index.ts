export { Transaction, ApolloCache } from './core/cache';
export { Cache } from './core/types/Cache';
export { DataProxy } from './core/types/DataProxy';

export {
  InMemoryCache,
  InMemoryCacheConfig,
} from './inmemory/inMemoryCache';
export { defaultDataIdFromObject } from './inmemory/policies';
export * from './inmemory/types';
