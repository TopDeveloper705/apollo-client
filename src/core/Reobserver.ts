import { WatchQueryOptions } from './watchQueryOptions';
import { NetworkStatus } from './networkStatus';
import { ApolloQueryResult } from './types';
import { Observer } from '../utilities/observables/Observable';
import { Concast } from '../utilities/observables/Concast';
import { invariant } from 'ts-invariant';

export class Reobserver<TData, TVars> {
  constructor(
    private observer: Observer<ApolloQueryResult<TData>>,
    private options: WatchQueryOptions<TVars>,
    private fetch: (
      options: WatchQueryOptions<TVars>,
      newNetworkStatus?: NetworkStatus,
    ) => Concast<ApolloQueryResult<TData>>,
    private shouldFetch: (() => boolean) | false,
  ) {}

  private concast?: Concast<ApolloQueryResult<TData>>;

  public reobserve(
    newOptions?: Partial<WatchQueryOptions<TVars>>,
    newNetworkStatus?: NetworkStatus,
  ): Promise<ApolloQueryResult<TData>> {
    if (newOptions) {
      this.updateOptions(newOptions);
    } else {
      // When we call this.updatePolling(newOptions) in the branch above,
      // it takes care of calling this.updatePolling(). In this branch, we
      // still need to update polling, even if there were no newOptions.
      this.updatePolling();
    }

    const concast = this.fetch(this.options, newNetworkStatus);

    if (this.concast) {
      // We use the {add,remove}Observer methods directly to avoid
      // wrapping observer with an unnecessary SubscriptionObserver
      // object, in part so that we can remove it here without triggering
      // any unsubscriptions, because we just want to ignore the old
      // observable, not prematurely shut it down, since other consumers
      // may be awaiting this.concast.promise.
      this.concast.removeObserver(this.observer, true);
    }

    concast.addObserver(this.observer);

    return (this.concast = concast).promise;
  }

  public updateOptions(newOptions: Partial<WatchQueryOptions<TVars>>) {
    Object.keys(newOptions).forEach(key => {
      const value = (newOptions as any)[key];
      if (value !== void 0) {
        (this.options as any)[key] = value;
      }
    });

    this.updatePolling();

    return this;
  }

  public stop() {
    if (this.concast) {
      this.concast.removeObserver(this.observer);
      delete this.concast;
    }

    if (this.pollingInfo) {
      clearTimeout(this.pollingInfo.timeout);
      this.options.pollInterval = 0;
      this.updatePolling();
    }
  }

  private pollingInfo?: {
    interval: number;
    timeout: NodeJS.Timeout;
  };

  // Turns polling on or off based on this.options.pollInterval.
  private updatePolling() {
    const {
      pollingInfo,
      options: {
        pollInterval,
      },
    } = this;

    if (!pollInterval) {
      if (pollingInfo) {
        clearTimeout(pollingInfo.timeout);
        delete this.pollingInfo;
      }
      return;
    }

    if (pollingInfo &&
        pollingInfo.interval === pollInterval) {
      return;
    }

    invariant(
      pollInterval,
      'Attempted to start a polling query without a polling interval.',
    );

    // Go no further if polling is disabled.
    if (this.shouldFetch === false) {
      return;
    }

    const info = pollingInfo || (
      this.pollingInfo = {} as Reobserver<TData, TVars>["pollingInfo"]
    )!;

    info.interval = pollInterval;

    const maybeFetch = () => {
      if (this.pollingInfo) {
        if (this.shouldFetch && this.shouldFetch()) {
          this.reobserve({
            fetchPolicy: "network-only",
          }, NetworkStatus.poll).then(poll, poll);
        } else {
          poll();
        }
      };
    };

    const poll = () => {
      const info = this.pollingInfo;
      if (info) {
        clearTimeout(info.timeout);
        info.timeout = setTimeout(maybeFetch, info.interval);
      }
    };

    poll();
  }
}
