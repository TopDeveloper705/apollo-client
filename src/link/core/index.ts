export * from './link';
export {
  createOperation,
  toPromise,
  fromPromise,
  fromError,
} from './linkUtils';
export * from './types';
