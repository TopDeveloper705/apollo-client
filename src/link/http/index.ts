export {
  ServerError,
  ServerParseError,
  ClientParseError,
  HttpQueryOptions,
  HttpConfig,
  UriFunction,
  Body,
  HttpOptions,
  fallbackHttpConfig,
  throwServerError,
  parseAndCheckHttpResponse,
  checkFetcher,
  createSignalIfSupported,
  selectHttpOptionsAndBody,
  serializeFetchParameter,
  selectURI
} from './common';

export {
  createHttpLink,
  HttpLink
} from './httpLink';
