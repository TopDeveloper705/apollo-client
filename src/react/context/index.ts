import { checkDEV } from "../../utilities";
checkDEV();

export * from './ApolloConsumer';
export * from './ApolloContext';
export * from './ApolloProvider';
