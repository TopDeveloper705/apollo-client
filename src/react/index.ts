export { __DEV__ } from "../utilities";

export {
  ApolloProvider,
  ApolloConsumer,
  getApolloContext,
  resetApolloContext,
  ApolloContextValue
} from './context';

export * from './hooks';

export {
  DocumentType,
  IDocumentDefinition,
  operationName,
  parser
} from './parser';

export * from './types/types';
