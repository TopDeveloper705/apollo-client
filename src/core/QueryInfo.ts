import { ExecutionResult, DocumentNode, GraphQLError } from 'graphql';
import { equal } from "@wry/equality";

import { Cache } from '../cache/core/types/Cache';
import { ApolloCache } from '../cache/core/cache';
import { WatchQueryOptions } from './watchQueryOptions';
import { ObservableQuery } from './ObservableQuery';
import { QueryListener } from './types';
import { ObservableSubscription } from '../utilities/observables/Observable';
import { isNonEmptyArray } from '../utilities/common/arrays';
import { graphQLResultHasError } from '../utilities/common/errorHandling';
import {
  NetworkStatus,
  isNetworkRequestInFlight,
} from './networkStatus';

export type QueryStoreValue = Pick<QueryInfo,
  | "variables"
  | "networkStatus"
  | "networkError"
  | "graphQLErrors"
  >;

// A QueryInfo object represents a single query managed by the
// QueryManager, which tracks all QueryInfo objects by queryId in its
// this.queries Map. QueryInfo objects store the latest results and errors
// for the given query, and are responsible for reporting those results to
// the corresponding ObservableQuery, via the QueryInfo.notify method.
// Results are reported asynchronously whenever setDirty marks the
// QueryInfo object as dirty, though a call to the QueryManager's
// broadcastQueries method may trigger the notification before it happens
// automatically. This class used to be a simple interface type without
// any field privacy or meaningful methods, which is why it still has so
// many public fields. The effort to lock down and simplify the QueryInfo
// interface is ongoing, and further improvements are welcome.
export class QueryInfo {
  listeners = new Set<QueryListener>();
  document: DocumentNode | null = null;
  lastRequestId = 1;
  subscriptions = new Set<ObservableSubscription>();
  variables?: Record<string, any>;
  networkStatus?: NetworkStatus;
  networkError?: Error | null;
  graphQLErrors?: ReadonlyArray<GraphQLError>;

  constructor(private cache: ApolloCache<any>) {}

  public init(query: {
    document: DocumentNode;
    variables: Record<string, any> | undefined,
    // The initial networkStatus for this fetch, most often
    // NetworkStatus.loading, but also possibly fetchMore, poll, refetch,
    // or setVariables.
    networkStatus?: NetworkStatus,
    observableQuery?: ObservableQuery<any>;
    lastRequestId?: number;
  }): this {
    let networkStatus = query.networkStatus || NetworkStatus.loading;
    if (this.variables &&
        this.networkStatus !== NetworkStatus.loading &&
        !equal(this.variables, query.variables)) {
      networkStatus = NetworkStatus.setVariables;
    }

    Object.assign(this, {
      document: query.document,
      variables: query.variables,
      networkError: null,
      graphQLErrors: this && this.graphQLErrors || [],
      networkStatus,
    });

    if (query.observableQuery) {
      this.setObservableQuery(query.observableQuery);
    }

    if (query.lastRequestId) {
      this.lastRequestId = query.lastRequestId;
    }

    return this;
  }

  private dirty: boolean = false;

  public setDirty(): this {
    if (!this.dirty) {
      this.dirty = true;
      if (!this.notifyTimeout) {
        this.notifyTimeout = setTimeout(() => this.notify(), 0);
      }
    }
    return this;
  }

  private notifyTimeout?: ReturnType<typeof setTimeout>;

  private diff: Cache.DiffResult<any> | null = null;

  setDiff(diff: Cache.DiffResult<any> | null) {
    const oldDiff = this.diff;
    this.diff = diff;
    if (!this.dirty && diff?.result !== oldDiff?.result) {
      this.setDirty();
    }
  }

  getDiff() {
    if (!this.diff) {
      const oq = this.observableQuery;
      const lastResult = oq && oq.getLastResult();
      const lastError = oq && oq.getLastError();
      const fetchPolicy = oq && oq.options.fetchPolicy || "cache-first";
      const errorPolicy = this.getErrorPolicy();
      const errorStatusChanged =
        errorPolicy !== 'none' &&
        (lastError && lastError.graphQLErrors) !== this.graphQLErrors;

      if (lastResult && lastResult.data && !errorStatusChanged) {
        this.diff = {
          result: lastResult.data,
          complete: true,
        };
      } else if (fetchPolicy !== "no-cache" &&
                 fetchPolicy !== "network-only") {
        this.diff = this.cache.diff({
          query: this.document as DocumentNode,
          variables: this.variables,
          returnPartialData: true,
          optimistic: true,
        });
      }
    }

    return this.diff!;
  }

  public readonly observableQuery: ObservableQuery<any> | null = null;
  private oqListener?: QueryListener;

  setObservableQuery(oq: ObservableQuery<any> | null) {
    if (oq === this.observableQuery) return;

    if (this.oqListener) {
      this.listeners.delete(this.oqListener);
    }

    (this as any).observableQuery = oq;

    if (oq) {
      this.listeners.add(this.oqListener = () => oq.reobserve());
    } else {
      delete this.oqListener;
    }
  }

  private getErrorPolicy() {
    const oq = this.observableQuery;
    return oq && oq.options.errorPolicy || "none";
  }

  notify() {
    if (this.notifyTimeout) {
      clearTimeout(this.notifyTimeout);
      this.notifyTimeout = void 0;
    }

    if (this.shouldNotify() && this.getDiff()) {
      this.listeners.forEach(listener => listener(this));
    }

    this.dirty = false;
  }

  private shouldNotify() {
    if (!this.dirty || !this.listeners.size) {
      return false;
    }

    if (!this.observableQuery) {
      return true;
    }

    const {
      fetchPolicy,
      returnPartialData,
      notifyOnNetworkStatusChange,
    } = this.observableQuery.options;

    if (fetchPolicy === "standby") {
      return false;
    }

    if (isNetworkRequestInFlight(this.networkStatus)) {
      const lastResult = this.observableQuery.getLastResult();

      const networkStatusChanged = !!(
        lastResult &&
        lastResult.networkStatus !== this.networkStatus
      );

      const shouldNotifyIfLoading =
        returnPartialData ||
        this.networkStatus === NetworkStatus.setVariables ||
        (networkStatusChanged && notifyOnNetworkStatusChange) ||
        fetchPolicy === 'cache-only' ||
        fetchPolicy === 'cache-and-network';

      if (!shouldNotifyIfLoading) {
        return false;
      }
    }

    return true;
  }

  public stop() {
    this.cancel();
    // Revert back to the no-op version of cancel inherited from
    // QueryInfo.prototype.
    delete this.cancel;

    this.variables =
    this.networkStatus =
    this.networkError =
    this.graphQLErrors = void 0;
  }

  // This method can be overridden for a given instance.
  public cancel() {}

  public updateWatch(options: WatchQueryOptions): this {
    // TODO Always cancelling here may be bad because it disrupts
    // cache.watch continuity.
    this.cancel();

    this.cancel = this.cache.watch({
      query: this.document!,
      variables: options.variables,
      optimistic: true,
      callback: diff => this.setDiff(diff),
    });

    return this;
  }

  public markResult<T>(
    result: ExecutionResult<T>,
    { variables,
      fetchPolicy,
      errorPolicy,
    }: Pick<WatchQueryOptions,
      | "variables"
      | "fetchPolicy"
      | "errorPolicy"
      >,
    allowCacheWrite: boolean,
  ) {
    if (fetchPolicy === 'no-cache') {
      this.diff = { result: result.data, complete: true };

    } else if (allowCacheWrite) {
      const ignoreErrors = errorPolicy === 'ignore' || errorPolicy === 'all';
      let writeWithErrors = !graphQLResultHasError(result);
      if (!writeWithErrors && ignoreErrors && result.data) {
        writeWithErrors = true;
      }

      if (writeWithErrors) {
        // Using a transaction here so we have a chance to read the result
        // back from the cache before the watch callback fires as a result
        // of writeQuery, so we can store the new diff quietly and ignore
        // it when we receive it redundantly from the watch callback.
        this.cache.performTransaction(cache => {
          cache.writeQuery({
            query: this.document!,
            data: result.data as T,
            variables,
          });

          const diff = cache.diff<T>({
            query: this.document!,
            variables,
            returnPartialData: true,
            optimistic: true,
          });

          // If we're allowed to write to the cache, and we can read a
          // complete result from the cache, update result.data to be the
          // result from the cache, rather than the raw network result.
          // Set without setDiff to avoid triggering a notify call, since
          // we have other ways of notifying for this result.
          this.diff = diff;
          if (diff.complete) {
            result.data = diff.result;
          }
        });
      }
    }

    this.graphQLErrors = isNonEmptyArray(result.errors) ? result.errors : [];
  }

  public markReady() {
    this.networkError = null;
    return this.networkStatus = NetworkStatus.ready;
  }

  public markError(error: Error) {
    this.networkStatus = NetworkStatus.error;
    return this.networkError = error;
  }
}
