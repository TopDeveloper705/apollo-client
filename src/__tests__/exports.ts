import * as client from "..";
import * as core from "../core";
import * as react from "../react";
import * as cache from "../cache";
import * as utilities from "../utilities";
import * as ssr from "../react/ssr";
import * as components from "../react/components";
import * as hoc from "../react/hoc";
import * as batch from "../link/batch";
import * as batchHttp from "../link/batch-http";
import * as context from "../link/context";
import * as error from "../link/error";
import * as retry from "../link/retry";
import * as schema from "../link/schema";
import * as ws from "../link/ws";
import * as http from "../link/http";

type Namespace = object;

describe('exports of public entry points', () => {
  function check(id: string, ns: Namespace) {
    it(id, () => {
      expect(Object.keys(ns).sort()).toMatchSnapshot();
    });
  }

  check("@apollo/client", client);
  check("@apollo/client/cache", cache);
  check("@apollo/client/core", core);
  check("@apollo/client/link/batch", batch);
  check("@apollo/client/link/batch-http", batchHttp);
  check("@apollo/client/link/context", context);
  check("@apollo/client/link/error", error);
  check("@apollo/client/link/http", http);
  check("@apollo/client/link/retry", retry);
  check("@apollo/client/link/schema", schema);
  check("@apollo/client/link/ws", ws);
  check("@apollo/client/react", react);
  check("@apollo/client/react/components", components);
  check("@apollo/client/react/hoc", hoc);
  check("@apollo/client/react/ssr", ssr);
  check("@apollo/client/utilities", utilities);
});
