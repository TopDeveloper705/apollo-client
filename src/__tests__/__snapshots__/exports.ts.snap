// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exports of public entry points @apollo/client 1`] = `
Array [
  "ApolloCache",
  "ApolloClient",
  "ApolloConsumer",
  "ApolloError",
  "ApolloLink",
  "ApolloProvider",
  "Cache",
  "DocumentType",
  "HttpLink",
  "InMemoryCache",
  "MissingFieldError",
  "NetworkStatus",
  "Observable",
  "ObservableQuery",
  "checkFetcher",
  "concat",
  "createHttpLink",
  "createSignalIfSupported",
  "defaultDataIdFromObject",
  "disableExperimentalFragmentVariables",
  "disableFragmentWarnings",
  "empty",
  "enableExperimentalFragmentVariables",
  "execute",
  "fallbackHttpConfig",
  "from",
  "fromError",
  "fromPromise",
  "getApolloContext",
  "gql",
  "isApolloError",
  "isReference",
  "makeReference",
  "makeVar",
  "operationName",
  "parseAndCheckHttpResponse",
  "parser",
  "resetApolloContext",
  "resetCaches",
  "rewriteURIForGET",
  "selectHttpOptionsAndBody",
  "selectURI",
  "serializeFetchParameter",
  "split",
  "throwServerError",
  "toPromise",
  "useApolloClient",
  "useLazyQuery",
  "useMutation",
  "useQuery",
  "useSubscription",
]
`;

exports[`exports of public entry points @apollo/client/cache 1`] = `
Array [
  "ApolloCache",
  "Cache",
  "InMemoryCache",
  "MissingFieldError",
  "cacheSlot",
  "defaultDataIdFromObject",
  "isReference",
  "makeReference",
  "makeVar",
]
`;

exports[`exports of public entry points @apollo/client/core 1`] = `
Array [
  "ApolloCache",
  "ApolloClient",
  "ApolloError",
  "ApolloLink",
  "Cache",
  "HttpLink",
  "InMemoryCache",
  "MissingFieldError",
  "NetworkStatus",
  "Observable",
  "ObservableQuery",
  "checkFetcher",
  "concat",
  "createHttpLink",
  "createSignalIfSupported",
  "defaultDataIdFromObject",
  "disableExperimentalFragmentVariables",
  "disableFragmentWarnings",
  "empty",
  "enableExperimentalFragmentVariables",
  "execute",
  "fallbackHttpConfig",
  "from",
  "fromError",
  "fromPromise",
  "gql",
  "isApolloError",
  "isReference",
  "makeReference",
  "makeVar",
  "parseAndCheckHttpResponse",
  "resetCaches",
  "rewriteURIForGET",
  "selectHttpOptionsAndBody",
  "selectURI",
  "serializeFetchParameter",
  "split",
  "throwServerError",
  "toPromise",
]
`;

exports[`exports of public entry points @apollo/client/link/batch 1`] = `
Array [
  "BatchLink",
  "OperationBatcher",
]
`;

exports[`exports of public entry points @apollo/client/link/batch-http 1`] = `
Array [
  "BatchHttpLink",
]
`;

exports[`exports of public entry points @apollo/client/link/context 1`] = `
Array [
  "setContext",
]
`;

exports[`exports of public entry points @apollo/client/link/error 1`] = `
Array [
  "ErrorLink",
  "onError",
]
`;

exports[`exports of public entry points @apollo/client/link/http 1`] = `
Array [
  "HttpLink",
  "checkFetcher",
  "createHttpLink",
  "createSignalIfSupported",
  "fallbackHttpConfig",
  "parseAndCheckHttpResponse",
  "rewriteURIForGET",
  "selectHttpOptionsAndBody",
  "selectURI",
  "serializeFetchParameter",
]
`;

exports[`exports of public entry points @apollo/client/link/retry 1`] = `
Array [
  "RetryLink",
]
`;

exports[`exports of public entry points @apollo/client/link/schema 1`] = `
Array [
  "SchemaLink",
]
`;

exports[`exports of public entry points @apollo/client/link/ws 1`] = `
Array [
  "WebSocketLink",
]
`;

exports[`exports of public entry points @apollo/client/react 1`] = `
Array [
  "ApolloConsumer",
  "ApolloProvider",
  "DocumentType",
  "getApolloContext",
  "operationName",
  "parser",
  "resetApolloContext",
  "useApolloClient",
  "useLazyQuery",
  "useMutation",
  "useQuery",
  "useSubscription",
]
`;

exports[`exports of public entry points @apollo/client/react/components 1`] = `
Array [
  "Mutation",
  "Query",
  "Subscription",
]
`;

exports[`exports of public entry points @apollo/client/react/hoc 1`] = `
Array [
  "graphql",
  "withApollo",
  "withMutation",
  "withQuery",
  "withSubscription",
]
`;

exports[`exports of public entry points @apollo/client/react/ssr 1`] = `
Array [
  "RenderPromises",
  "getDataFromTree",
  "getMarkupFromTree",
  "renderToStringWithData",
]
`;

exports[`exports of public entry points @apollo/client/utilities 1`] = `
Array [
  "Concast",
  "DeepMerger",
  "Observable",
  "addTypenameToDocument",
  "argumentsObjectFromField",
  "asyncMap",
  "buildQueryFromSelectionSet",
  "canUseWeakMap",
  "checkDocument",
  "cloneDeep",
  "concatPagination",
  "createFragmentMap",
  "getDefaultValues",
  "getDirectiveNames",
  "getFragmentDefinition",
  "getFragmentDefinitions",
  "getFragmentFromSelection",
  "getFragmentQueryDocument",
  "getInclusionDirectives",
  "getMainDefinition",
  "getOperationDefinition",
  "getOperationName",
  "getQueryDefinition",
  "getStoreKeyName",
  "getTypenameFromResult",
  "graphQLResultHasError",
  "hasClientExports",
  "hasDirectives",
  "isField",
  "isInlineFragment",
  "isNonEmptyArray",
  "isReference",
  "iterateObserversSafely",
  "makeReference",
  "maybeDeepFreeze",
  "mergeDeep",
  "mergeDeepArray",
  "offsetLimitPagination",
  "relayStylePagination",
  "removeArgumentsFromDocument",
  "removeClientSetsFromDocument",
  "removeConnectionDirectiveFromDocument",
  "removeDirectivesFromDocument",
  "removeFragmentSpreadFromDocument",
  "resultKeyNameFromField",
  "shouldInclude",
  "storeKeyNameFromField",
  "valueToObjectRepresentation",
]
`;
